### ISSUE TITLE
Hangs on ceph-mon : collect admin and bootstrap keys

### ISSUE THREAD

**Issue by tbenst** [ceph/ceph-ansible#760]
I am attempting to setup Ceph Jewel via Ceph Ansible on Debian Jessie using backports. ceph-common deploys fine but I got stuck on ceph-mon : collect admin and bootstrap keys.

I was able to resolve the issue by running the playbook with cephx set to false, and then running again with cephx set to true.

**Comment by leseb** [comment]
@tbenst did the monitor start? If we can't collect keys it usually means the monitor didn't start.

**Comment by tbenst** [comment]
Correct, the monitor did not start

**Comment by leseb** [comment]
I think we are having the same issue on RedHat systems, I'm investigating at the moment.

**Comment by leseb** [comment]
@tbenst can you check again from master? I believe #764 fixes it.

**Comment by tbenst** [comment]
Looks good, thanks!!

**Comment by leseb** [comment]
cool!

**Comment by maxtortime** [comment]
I have s still this problem when I use a ubuntu wily release.
When first provision, I had showed some 404 error because monitor cannot find wily's ceph download's directory.

But 2nd provision, I hadn't shown that error, but just hangs on ceph-mon..

**Comment by leseb** [comment]
@maxtortime if you have an issue with package install this is not ceph-ansible's fault, it is more a platform hiccup.
Where does it hang now?

**Comment by maxtortime** [comment]
I renewed my reply to #818 .

**Comment by mclamb** [comment]
I am experiencing this issue as well: they playbook locks up during ceph-mon: collect admin and bootstrap keys. I am using ceph-ansible master, deploying Jewel on Ubuntu 14.04 targets. It appears as though it is hanging on a newly spawned instance of ceph-create-keys, which has slightly different parameter stylings than the instance of ceph-create keys that seemed to be started by ceph-mon? e.g.:

Ssl  23:11   0:47 /usr/bin/ceph-mon --cluster=ceph -i test-storage1 -f --setuser ceph --setgroup ceph
Ss   23:11   0:01 /usr/bin/python /usr/sbin/ceph-create-keys --cluster=ceph -i test-storage1
S+   23:23   0:00 /usr/bin/python /usr/sbin/ceph-create-keys --cluster ceph --id test-storage1

**Comment by leseb** [comment]
No `monitor_interface` and `public_network` don't need to be on the same LAN, however the `public_network` should be accessible from all the mons.

**Comment by mclamb** [comment]
It turns out I *think* the issue was due to the fact that ceph-mon was still running when I was redeploying with configuration changes. If I `sudo stop ceph-all` on all nodes prior to running the playbook, it appears to get beyond that task and finish successfully. The playbook does not appear to restart osds on those nodes (I am colocating mons and osds in this case, which I know is not recommended, but I hope isn't too terrible?) so I then go back and `sudo start ceph-all` and I get a healthy cluster.

**Comment by dang** [comment]
I'm seeing this on a clean setup with Fedora 23 as the host.  The mon never generates the admin key. I'm investigating, but this is definitely an issue.

**Comment by dang** [comment]
This appears to be related to running upstream dev (currently v11.0.0-2347-g11c9039).  Switching to stable does not exhibit this hang.

**Comment by batrick** [comment]
@leseb, I also had this problem. I think this fixes it: https://github.com/ceph/ceph/pull/11123

I'm going to give it a try once gitbuilders is done.

**Comment by batrick** [comment]
@leseb, I can confirm ceph/ceph#11123 fixes it.

**Comment by leseb** [comment]
Thanks @dang and @batrick for looking into this.
If https://github.com/ceph/ceph/pull/11123 fixes the issue, I think we can re-close this.

Let me know.

**Comment by mcv21** [comment]
I'm seeing this issue on trying to bootstrap a cluster with 3 mons ; ansible is hanging at "collect admin and bootstrap keys" and each mon is syslogging `sto-1-1 ceph-create-keys[31795]: INFO:ceph-create-keys:ceph-mon is not in quorum: u'probing'`
All the machines can reach each others mon addresses.
This is Ubuntu 16.04 (distro packages). Happy to provide more debugging if it helps.

**Comment by mclamb** [comment]
I am still running into this issue. I thought my previous comment about `sudo stop ceph-all` fixed it, but apparently not. I also have the latest ceph in which ceph-create-keys def_keys() is quite different from the the one mentioned in ceph/ceph#11123 so it does not appear to be the source of the problem.

The hang at "collect admin and bootstrap keys" eventually returns but it never created the client.admin key so the following task "wait for client.admin key exists" eventually times out and everything errs.

Any ideas? I am just deploying to a single mon and osd (colocated) within a VM for now.

Thanks.

**Comment by mclamb** [comment]
I see the problem: ceph-mon is listening on port 6789 on the IP of the `public_network` and not on the IP of the `monitor_interface`

When I run ceph-create-keys manually I can see that it is attempting to contact the mon on the wrong address.

In my group_vars/all I have `monitor_interface` set to the interface for one network, then both `public_network` and `private_network` are set for yet two other networks.

According to `/etc/ceph/ceph.conf` the `mon host = 10.0.20.31` which is the correct IP address of the `monitor_interface` and the `public network = 10.0.30/0/24` and `private network = 10.0.40/24`.

`ceph-mon` is listening only on the public network's 10.0.30.31:6789

Is this expected behavior? Seems as though I must set `monitor_interface` to be the same interface as `public_network` in order to get it to work.

Thanks.

**Comment by leseb** [comment]
I've never really encountered that. Usually we can have 3 different networks: mon, public and cluster. Unless you have a real need for isolation and want to separate public from monitor then you should use different subnet. IMHO keeping monitor and public network as part of the same subnet is fine.

**Comment by mclamb** [comment]
I presume ceph-mon should listen on the interface defined by `monitor_interface` right? In this case it seems like there might be a bug somewhere that is causing it to listen on the interface associated with the `public_network`.

**Comment by leseb** [comment]
@mclamb You're right, do you mind opening a new issue describing this?
Thanks!

**Comment by jiwhiz** [comment]
I had the same issue with v2.0 release. Here are the steps I did:

Clone the repo and switch to v2.0 release:
```
git clone https://github.com/ceph/ceph-ansible.git
cd ceph-ansible
git checkout v2.0
```

Create vagrant_variables.yml and site.yml
```
cp vagrant_variables.yml.sample vagrant_variables.yml
cp site.yml.sample site.yml
```

Set `ceph_stable` to true in Vagrantfile:
```
  ansible.extra_vars = {
      cluster_network: "#{CLUSTER_SUBNET}.0/24",
      journal_size: 100,
      public_network: "#{PUBLIC_SUBNET}.0/24",
      ceph_stable: 'true'
  }
```

Create VMs:
```
vagrant up --no-provision --provider=virtualbox
vagrant provision
```

Then it hangs at 
```
TASK [ceph-mon : collect admin and bootstrap keys] *****************************
```

I'm using Ansible 2.2.0.0 with Mac OS X.

"cephx set to false" did not work. Any other workaround?

**Comment by leseb** [comment]
@jiwhiz can you check the state of the monitor?

@mclamb we just pushed this patch in Ceph to fix this issue: https://github.com/ceph/ceph/pull/12188
